src/java/org/apache/cassandra/net/Verb.java

```diff
@@ -434,12 +459,80 @@ public static Verb fromId(int id)
+     *
+     * <p>Relative ids can be used for indexing idToCustomVerbMap. Absolute ids exist to distinguish
+     * regular verbs from custom verbs in the id space.</p>
+     *
+     * @param id the relative or absolute id.
+     * @return a relative id if {@code id} is absolute, or absolute id if {@code id} is relative.
      */
     private static int idForCustomVerb(int id)
     {
         return CUSTOM_VERB_START - id;
     }
+
+    /**
+     * Add a new custom verb to the list of verbs.
+     *
+     * <p>While we could dynamically generate an {@code id} for callers, it's safer to have users
+     * explicitly control the id space since it prevents nodes with different versions disagreeing on which
+     * verb has which id, e.g. during upgrade.</p>
+     *
+     * @param name the name of the new verb.
+     * @param id the identifier for this custom verb (must be >= 0 and <= MAX_CUSTOM_VERB_ID).
+     * @param priority the priority of the new verb.
+     * @param expiration an optional timeout for this verb. @see VerbTimeouts.
+     * @param stage The stage this verb should execute in.
+     * @param serializer A method to serialize this verb
+     * @param handler A method to handle this verb when received by the network
+     * @param responseVerb The verb to respond with (optional)
+     * @return A Verb for the newly added verb.
+     */
+    public static synchronized Verb addCustomVerb(String name, int id, Priority priority, ToLongFunction<TimeUnit> expiration, Stage stage, Supplier<? extends IVersionedAsymmetricSerializer<?, ?>> serializer, Supplier<? extends IVerbHandler<?>> handler, Verb responseVerb)
+    {
+        assertNameIsUnused(name);
+        assertCustomIdIsUnused(idToCustomVerbMap, id, name);
+        Verb verb =  new Verb(name, CUSTOM, id, priority, expiration, stage, serializer, handler, responseVerb);
+
+        int absoluteId = idForCustomVerb(id);
+        minCustomId = Math.min(absoluteId, minCustomId);
+
+        Verb[] newMap = Arrays.copyOf(idToCustomVerbMap, CUSTOM_VERB_START - minCustomId + 1);
+        System.arraycopy(idToCustomVerbMap, 0, newMap, 0, idToCustomVerbMap.length);
+        newMap[id] = verb;
+        idToCustomVerbMap = newMap;
+
+        return verb;
+    }
+
+    private static void assertNameIsUnused(String name)
+    {
+        if (verbs.stream().map(v -> v.name).collect(Collectors.toList()).contains(name))
+            throw new IllegalArgumentException("Verb name '" + name + "' already exists");
+    }
+
+    /**
+     * Decorates the specified verb handler with the provided method.
+     *
+     * <p>An example use case is to run a custom method after every write request.</p>
+     *
+     * @param verbs the list of verbs whose handlers should be wrapped by {@code decoratorFn}.
+     * @param decoratorFn the method that decorates the handlers in verbs.
+     */
+    public static void decorateHandler(List<Verb> verbs, Function<IVerbHandler<?>, IVerbHandler<?>> decoratorFn)
```

Comment:
this method is racy, is that ok? We might lose a decorator during concurrent invocations for the same verbs.

Reply:
Ouch. Great catch.